### PR TITLE
refactoring to change low level enumerator back to normal

### PR DESCRIPTION
**Pull request by xiongrenyi**
it was changed to low level enumerator API to add some perf metrics which actually can be done in normal enumerator as well.

### CODE REVIEW

**Comment by qualiu** [comment]
LGTM  (with following test)

```
test\csharp\testKeyValueStream\test-by-starting-socket.bat -p 9113 -e 10240 -n 200 -d 1 -t 6 2>&1 | lzmw -it "args.\d+|sizeof|final|end of|begin of|exception|OutOfMemory\w*|--\S+-memory" --nt "weak|InterruptedException" -e "elementCount|testTime\w*"

test\csharp\testKeyValueStream\test-by-starting-socket.bat -c chkDir -p 9116 -w 6 -s 1 -e 10240 -n 200 -d 1 -t 6 2>&1 | lzmw -it "args.\d+|sizeof|final|end of|begin of|exception|OutOfMemory\w*|--\S+-memory" --nt "weak|InterruptedException" -e "elementCount|testTime\w*"

test\tools\start-zookeeper-kafka.bat
test\csharp\kafkaStreamTest\test.bat WindowSlideTest -Topics test -d 1 -w 5 -s 1 2>&1 | lzmw -it "args.\d+|sumcount|exception"

test\csharp\kafkaStreamTest\create-2-topics-for-test.bat
test\csharp\kafkaStreamTest\test.bat UnionTopicTest -Topic1 id_count_1 -Topic2 id_count_2 -d 1 -f saveTxtDir -w 5 -s 1 2>&1 | lzmw -it "args.\d+|sumcount|exception"

::## this test need data , such as D:\cosmos\download-stream\tenant\csv-2015-10-01
test\csharp\txtStreamTest\test.bat D:\cosmos\download-stream\tenant\csv-2015-10-01  2>&1 | lzmw -it "exception|RDD count|used time|will read|submit" -e "\d+(\.\d+)?\s*m?s|true|false|rio\w*|x64"

```

**Comment by qintao1976** [comment]
LGTM

File: csharp/Worker/Microsoft.Spark.CSharp/Worker.cs [hunk 1]
```diff
@@ -455,185 +435,91 @@ private static long ToUnixTime(DateTime dt)
...
-                        else
-                        {
-                            result[0] = null;
-                        }
-                        break;
-                    }
 
-                case SerializedMode.Row:
-                    {
-                        Debug.Assert(messageLength > 0);
```
Comment (danielli90) [inline csharp/Worker/Microsoft.Spark.CSharp/Worker.cs]: I think we should keep Debug.Assert() here, to ensure the logging/assertion characteristics as the same as before, unless we deem this assert() is completely unnecessary.

File: csharp/Worker/Microsoft.Spark.CSharp/Worker.cs [hunk 2]
```diff
@@ -455,185 +435,91 @@ private static long ToUnixTime(DateTime dt)
...
+                    switch (mode)
                     {
-                        result = new object[1];
-                        if (messageLength > 0)
-                        {
-                            byte[] buffer = SerDe.ReadBytes(inputStream, messageLength);
-                            if (buffer == null)
+                        case SerializedMode.String:
                             {
-                                logger.LogDebug("Buffer is null. Message length is {0}", messageLength);
```
Comment (danielli90) [inline csharp/Worker/Microsoft.Spark.CSharp/Worker.cs]: I think we should keep this logging message, to ensure the logging/assertion characteristics as the same as before, unless we deem this logging line is completely unnecessary.

**Comment by danielli90** [comment]
LGTM, except two minor inline comments suggesting to keep the logging/assertion characteristics as the same as before, unless this specific logging line and assertion are deemed completely unnecessary.

**Comment by xiongrenyi** [comment]
@danielli90  both make sense, added back.

**Comment by hebinhuang** [comment]
LGTM

**Pull request merged**